### PR TITLE
Add tool to generate recommended config

### PR DESCRIPTION
**Pull request by armano2**
#144

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -12,13 +12,14 @@
   "main": "lib/index.js",
   "scripts": {
     "lint": "eslint lib/ tests/",
```
Comment (weirdpattern) [inline package.json]: Not sure if this was left out on purpose or not...
Do we want to include `tools/` here too?

Comment (armano2) [reply package.json]: ahh i missed it 👍

**Comment by weirdpattern** [comment]
This looks great! 
Not so sure how I feel about the existing rules that are marked as recommended... but that can be discussed later

**Comment by armano2** [comment]
@weirdpattern that was not a goal of this PR, in each rule we should specify if it recommended or not.

**Comment by j-f1** [comment]
If you rename `recommended.js` to `recommended.json`, you won’t have to wrap it in the wrapper code you have.

**Comment by armano2** [comment]
but i'm not going to be able to put comment, but sure, i can change that

**Comment by j-f1** [comment]
What if it wasn’t checked into Git and was regenerated as necessary? That would make it clearer that it’s generated.

**Comment by armano2** [comment]
@j-f1 its better if file is always present for reference (it's easier to check what is enabled by it)

**Comment by weirdpattern** [comment]
Hey @j-f1 any way you can modify eslint-docs to read the recommended rules and list them in the README.md file?

**Comment by j-f1** [comment]
I’ll try to look at it tomorrow, but feel free to open a PR!

**Comment by weirdpattern** [comment]
I will take a look at it tonight

**Comment by armano2** [comment]
@weirdpattern we had something like that in [eslint-plugin-vue](https://github.com/vuejs/eslint-plugin-vue/tree/4.x)

File: tools/update-recommended.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+const fs = require("fs");
+
+/**
+ * Generate recommended configuration
+ * @returns {void}
+ */
+function generate() {
+    const rulesDir = path.resolve(__dirname, "../lib/rules");
+    const rules = fs
+        .readdirSync(rulesDir)
```
Comment (bradzacher) [inline tools/update-recommended.js]: Could you use `requireIndex` instead of manually writing it?

File: package.json [hunk 3]
```diff
@@ -11,14 +11,15 @@
...
+    "lint": "eslint lib/ tests/ tools/",
+    "lint:fix": "eslint lib/ tests/ tools/ --fix",
     "docs": "eslint-docs",
     "docs:check": "eslint-docs check",
     "format-no-write": "prettier-eslint lib/**/*.js tests/**/*.js --eslint-config-path=.eslintrc --eslint-ignore --prettier-ignore --eslint-path=node_modules/eslint --config=.prettierrc",
     "format": "yarn format-no-write --write",
     "format-check": "yarn format-no-write --list-different",
-    "test": "mocha tests --recursive --reporter=dot"
+    "test": "mocha tests --recursive --reporter=dot",
+    "update": "node tools/update-recommended.js"
```
Comment (bradzacher) [inline package.json]: could you please rename this to `recommended:update` just so it's more descriptive?

File: package.json [hunk 4]
```diff
@@ -11,14 +11,15 @@
   "author": "Nicholas C. Zakas",
   "main": "lib/index.js",
   "scripts": {
-    "lint": "eslint lib/ tests/",
-    "lint:fix": "eslint lib/ tests/ --fix",
+    "lint": "eslint lib/ tests/ tools/",
+    "lint:fix": "eslint lib/ tests/ tools/ --fix",
     "docs": "eslint-docs",
     "docs:check": "eslint-docs check",
     "format-no-write": "prettier-eslint lib/**/*.js tests/**/*.js --eslint-config-path=.eslintrc --eslint-ignore --prettier-ignore --eslint-path=node_modules/eslint --config=.prettierrc",
```
Comment (bradzacher) [inline package.json]: I missed this - add `tests/**/*.js` to this script too

Comment (armano2) [reply package.json]: @bradzacher ohh i forgot about this one to, 

added

**Comment by bradzacher** [comment]
OOOOOKAY

I went through 4 configs that I know with config for our plugin, and came up with the following recommendations by merging them together.

Please provide feedback.


## The configs for reference:
<details><summary>eslint-config-typescript-recommended</summary>
<p>

https://github.com/diva-e/eslint-config-typescript-recommended/blob/master/index.js#L124-L140

```JSON
{
    "typescript/adjacent-overload-signatures": "error",
    "typescript/class-name-casing": "error",
    "typescript/explicit-member-accessibility": "error",
    "typescript/interface-name-prefix": ["error", "always"],
    "typescript/member-ordering": "error",
    "typescript/no-angle-bracket-type-assertion": "error",
    "typescript/no-empty-interface": "error",
    "typescript/no-explicit-any": "off",
    "typescript/no-namespace": "error",
    "typescript/no-parameter-properties": "off",
    "typescript/no-triple-slash-reference": "error",
    "typescript/no-type-alias": "error",
    "typescript/no-var-requires": "error",
    "typescript/prefer-namespace-keyword": "error"
}
```

</p>
</details>

<details><summary>eslint-config-typescript</summary>
<p>

https://github.com/weirdpattern/eslint-config-typescript/blob/master/src/typescript.js#L21-L68

```JSON
{
    "typescript/adjacent-overload-signatures": "error",
    "typescript/class-name-casing": "error",
    "typescript/explicit-function-return-type": [
        "error",
        {
            "allowExpressions": true
        }
    ],
    "typescript/explicit-member-accessibility": "error",
    "typescript/interface-name-prefix": "error",
    "typescript/member-delimiter-style": "error",
    "typescript/member-ordering": [
        "error",
        {
            "default": [
                "static-field",
                "private-field",
                "protected-field",
                "public-field",
                "constructor",
                "public-method",
                "protected-method",
                "private-method"
            ]
        }
    ],
    "typescript/no-angle-bracket-type-assertion": "error",
    "typescript/no-array-constructor": "error",
    "typescript/no-inferrable-types": [
        "error",
        {
            "ignoreProperties": true,
            "ignoreParameters": true
        }
    ],
    "typescript/no-namespace": [
        "error",
        {
            "allowDeclarations": true
        }
    ],
    "typescript/no-non-null-assertion": "error",
    "typescript/no-parameter-properties": "error",
    "typescript/no-triple-slash-reference": "error",
    "typescript/no-use-before-define": "error",
    "typescript/no-var-requires": "error",
    "typescript/prefer-namespace-keyword": "error",
    "typescript/type-annotation-spacing": "error"
}
```

</p>
</details>

<details><summary>eslint-config-standard-with-typescript</summary>
<p>

https://github.com/mightyiam/eslint-config-standard-with-typescript/blob/master/src/index.ts#L19-L32

```JSON
{
    "typescript/adjacent-overload-signatures": "error",
    "typescript/explicit-function-return-type": "error",
    "typescript/explicit-member-accessibility": "error",
    "typescript/member-delimiter-style": [
        "error",
        {
            "delimiter": "none"
        }
    ],
    "typescript/no-angle-bracket-type-assertion": "error",
    "typescript/no-array-constructor": "error",
    "typescript/no-empty-interface": "error",
    "typescript/no-namespace": "error",
    "typescript/no-non-null-assertion": "error",
    "typescript/no-triple-slash-reference": "error",
    "typescript/no-unused-vars": "error",
    "typescript/no-use-before-define": [
        "error",
        {
            "functions": false,
            "classes": false,
            "variables": false,
            "typedefs": false
        }
    ],
    "typescript/no-var-requires": "error",
    "typescript/type-annotation-spacing": "error"
}

```

</p>
</details>

<details><summary>eslint-config-assignar</summary>
<p>

This is the last company I worked at, and I did the eslint config :)
https://github.com/assignar/eslint-config-assignar/blob/develop/packages/eslint-config-assignar-base/src/plugins/typescript.ts

```JSON
{
    "typescript/adjacent-overload-signatures": "error",
    "typescript/class-name-casing": "error",
    "typescript/explicit-function-return-type": "off",
    "typescript/explicit-member-accessibility": "error",
    "typescript/interface-name-prefix": ["warn", "never"],
    "typescript/member-delimiter-style": [
        "error",
        {
            "delimiter": "none",
            "requireLast": true,
            "ignoreSingleLine": true
        }
    ],
    "typescript/member-naming": "off",
    "typescript/member-ordering": "off",
    "typescript/no-angle-bracket-type-assertion": "error",
    "typescript/no-array-constructor": "error",
    "typescript/no-empty-interface": "off",
    "typescript/no-explicit-any": "warn",
    "typescript/no-inferrable-types": "off",
    "typescript/no-namespace": [
        "error",
        {
            "allowDeclarations": false,
            "allowDefinitionFiles": true
        }
    ],
    "typescript/no-non-null-assertion": "warn",
    "typescript/no-parameter-properties": "error",
    "typescript/no-triple-slash-reference": "warn",
    "typescript/no-type-alias": "off",
    "typescript/no-unused-vars": "error",
    "typescript/no-use-before-define": "error",
    "typescript/no-var-requires": "off",
    "typescript/prefer-namespace-keyword": "off",
    "typescript/type-annotation-spacing": [
        "error",
        {
            "before": false,
            "after": true,
            "overrides": {
                "arrow": {
                    "before": true,
                    "after": true
                },
                "colon": {
                    "before": false,
                    "after": true
                }
            }
        }
    ]
}
```

</p>
</details>

## Recommended Config

```JSON
{
    "typescript/adjacent-overload-signatures": "error",
    "typescript/ban-types": [
        "error",
        {
            "types": {
                "Boolean": {
                    "message": "Use boolean instead",
                    "fixWith": "boolean"
                },
                "Number": {
                    "message": "Use number instead",
                    "fixWith": "number"
                },
                "Object": {
                    "message": "Use Record<string, any> instead",
                    "fixWith": "Record<string, any>"
                },
                "Symbol": {
                    "message": "Use symbol instead",
                    "fixWith": "symbol"
                }
            }
        }
    ],
    "typescript/camelcase": [
        "error",
        {
            "allow": ["^UNSAFE_"],
            "ignoreDestructuring": false,
            "properties": "never"
        }
    ],
    "typescript/class-name-casing": "error",
    "typescript/explicit-function-return-type": [
        "error",
        {
            "allowExpressions": true
        }
    ],
    "typescript/explicit-member-accessibility": "error",
    "typescript/generic-type-naming": ["error", "'^T[A-Z][a-zA-Z]+$'"],
    "typescript/interface-name-prefix": ["error", "never"],
    "typescript/member-delimiter-style": "error",
    "typescript/member-ordering": "error",
    "typescript/no-angle-bracket-type-assertion": "error",
    "typescript/no-array-constructor": "error",
    "typescript/no-empty-interface": "error",
    "typescript/no-explicit-any": "error",
    "typescript/no-inferrable-types": [
        "error",
        {
            "ignoreProperties": true,
            "ignoreParameters": true
        }
    ],
    "typescript/no-namespace": [
        "error",
        {
            "allowDeclarations": true
        }
    ],
    "typescript/no-non-null-assertion": "error",
    "typescript/no-parameter-properties": "error",
    "typescript/no-triple-slash-reference": "error",
    "typescript/no-type-alias": "off",
    "no-unused-vars": "off",
    "typescript/no-unused-vars": "off",
    "typescript/no-use-before-define": "error",
    "typescript/no-var-requires": "error",
    "typescript/prefer-namespace-keyword": "error",
    "typescript/type-annotation-spacing": "error"
}
```

**Comment by armano2** [comment]
Default configuration of plugin should be recommended one, than we should update default config for those rules.

**Comment by bradzacher** [comment]
`camelcase` is the general consensus I've seen for javascript code now-a-days. I have seen people disable properties due to the fact that APIs are often non-standard in their casing.

`class-name-casing` is another consensus I've seen from the community in regards javascript class names. The TS docs use PascalCase for all interfaces names from what I can see as well.

`generic-type-naming` is an open thing; it's a holdover from C# but most people don't have a preference I think.  
I haven't seen any consensus in the official docs about it (they switch between single character generic names T/K/U).
What do people think?

I'm happy to turn off `member-ordering`, I don't like it personally as well, but 2 of the configs had it on, so I turned it on...

**Comment by bradzacher** [comment]
okie dokie.

So I think the best course of action is to merge this PR first.
Then we'll go ahead and merge this.

And we'll then go about raising separate PR(s) to update the config to match the desired recommended set.

**Pull request merged**